amp/parse.go

```diff
@@ -32,6 +41,62 @@ type Size struct {
 	Width          int64
 }
 
+// Policy consent types
+const (
+	ConsentDenied = iota
+	ConsentTCF1
+	ConsentTCF2
+	ConsentUSPrivacy
+)
+
+func ReadPolicy(ampParams Params, req *openrtb2.BidRequest, pbsConfigGDPREnabled bool) (privacy.PolicyWriter, error) {
+
+	if len(ampParams.Consent) == 0 {
+		return privacy.NilPolicyWriter{}, nil
+	}
+
+	var rv privacy.PolicyWriter = privacy.NilPolicyWriter{}
+	var warning error
+	var errMsg string
+
+	switch ampParams.ConsentType {
+	case ConsentTCF2:
+		if pbsConfigGDPREnabled {
+			// Even if consent is invalid, we write the consent string to req.user.ext.consent
+			rv = gdpr.ConsentWriter{ampParams.Consent}
+
+			if tcf2.IsConsentV2(ampParams.Consent) {
+				if _, err := tcf2.ParseString(ampParams.Consent); err != nil {
+					errMsg = err.Error()
+				}
+			} else {
+				errMsg = fmt.Sprintf("Consent string '%s' is not a valid TCF2 consent string.", ampParams.Consent)
+			}
+		}
+	case ConsentUSPrivacy:
+		if ccpa.ValidateConsent(ampParams.Consent) {
+			rv = ccpa.ConsentWriter{ampParams.Consent}
+		} else {
+			errMsg = fmt.Sprintf("Consent string '%s' is not a valid CCPA consent string.", ampParams.Consent)
+		}
+	case ConsentDenied:
+		errMsg = "Consent denied. Consent string ignored."
+	case ConsentTCF1:
+		errMsg = "TCF1 consent is deprecated and no longer supported."
+	default:
+		errMsg = fmt.Sprintf("Consent '%d' is not recognized as either CCPA or GDPR TCF2.", ampParams.ConsentType)
```

Comment:
If there is consent type specified, we should fall back to the same logic of trying to parse it as TCF2 and CCPA. We don't want to just return an error message.

Reply:
Do you mean if there is no consent type specified? If the consent type is specified, then it should be an error if a different type of consent string is provided.

Reply:
Great catch. Implemented.

Reply:
@bsardo the logic has been updated. Let me know your thoughts.